shap/explainers/_tree.py

```diff
@@ -565,6 +565,19 @@ def shap_interaction_values(self, X, y=None, tree_limit=None):
             else:
                 self.expected_value = phi[0, -1, -1]
                 return phi[:, :-1, :-1]
+        elif (self.model.model_type == "catboost") and (self.feature_perturbation == "tree_path_dependent"): # thanks again to the CatBoost team for implementing this...
+            assert tree_limit == -1, "tree_limit is not yet supported for CatBoost models!"
+            import catboost
+            if type(X) != catboost.Pool:
+                X = catboost.Pool(X, cat_features=self.model.cat_feature_indices)
+            phi = self.model.original_model.get_feature_importance(data=X, fstr_type='ShapInteractionValues')
+            # note we pull off the last column and keep it as our expected_value
+            if len(phi.shape) == 4:
+                self.expected_value = getattr(self, "expected_value", [phi[0, i, -1, -1] for i in range(phi.shape[1])])
+                return [phi[:, i, :-1, :-1] for i in range(phi.shape[1])]
+            else:
+                self.expected_value = getattr(self, "expected_value", phi[0, -1, -1])
+                return phi[:, :-1, :-1]
```

Comment:
Do you think this diff should be it's own private method? e.g. 

```python
elif (self.model.model_type == "catboost") and (self.feature_perturbation == "tree_path_dependent"):
    assert tree_limit == -1, "tree_limit is not yet supported for CatBoost models!"
    return self._shap_interaction_values_catboost(X)
```

As a rather general observation, the shap codebase and particularly the tree explainer has some rather long functions with several chunky if/else statements. So, within reason we could look to identify blocks of code that could be broken down a little.

Reply:
I absolutely agree that these if/else statements make the functions clunky. But I would suggest that we gather a couple more tests and then tackle a general rewrite of the TreeExplainer.shap_values and TreeExplainer.shap_interaction_values methods (and some methods called in there, e.g. generate output, etc.). I strongly believe that there is a way better formulation than the one we have right now. 

So as for now, I would just leave it as it is.